---
title: "ChIPseq In Bioconductor part3"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_ChIPseq/"
output: 
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


# Data

In todays session we will continue to review the Myc ChIPseq we were working on in our last sessions.

This include Myc ChIPseq for MEL and Ch12 celllines.

Information and files for the [Myc ChIPseq in MEL cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000EUA/)

Information and files for the [Myc ChIPseq in Ch12 cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000ERN/)

<!-- --- -->

<!-- # Data -->

<!-- We will be working with peak calls today, so we can download the MACS2 peak calls from the Encode website. -->

<!-- [Myc Mel Rep1](https://www.encodeproject.org/files/ENCFF363WUG/@@download/ENCFF363WUG.bed.gz) -->

<!-- [Myc Mel Rep2](https://www.encodeproject.org/files/ENCFF139JHS/@@download/ENCFF139JHS.bed.gz) -->

<!-- [Myc Ch12 Rep1](https://www.encodeproject.org/files/ENCFF160KXR/@@download/ENCFF160KXR.bed.gz) -->

<!-- [Myc Ch12 Rep2](https://www.encodeproject.org/files/ENCFF962BGJ/@@download/ENCFF962BGJ.bed.gz) -->

---

## Data

I have provided peakcalls from MACS2 following the processing steps outlined in our last session.

Peak calls for Myc in MEL and Ch12 cellines can be found in 

**data/peaks/**

* **data/peaks/Mel_1_peaks.xls**
* **data/peaks/Mel_2_peaks.xls**
* **data/peaks/Ch12_1_peaks.xls**
* **data/peaks/Ch12_1_peaks.xls**

---
## Annotation of peaks to genes

In our last session we reviewed how we can [identify putative transciption factor binding sites using peak calling programs such as MACS2.](https://rockefelleruniversity.github.io/RU_ChIPseq/chipseq_course//presentations/slides/ChIPseq_In_Bioconductor2.html#45).

Since transcription factors, as suggested in name, may regulate the transcription of their target genes, we used the **ChIPseeker package** to associate our peaks, representing potential transcription factor binding events, to their overlapping or closest mm10 genes.

```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE}
library(GenomicRanges)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(ChIPseeker)
macsPeaks <- "data/peaks/Mel_1_peaks.xls"
macsPeaks_DF <- read.delim(macsPeaks,comment.char="#")
macsPeaks_GR <- GRanges(
 seqnames=macsPeaks_DF[,"chr"],
 IRanges(macsPeaks_DF[,"start"],macsPeaks_DF[,"end"])
)
mcols(macsPeaks_GR) <- macsPeaks_DF[,c("abs_summit", "fold_enrichment")]
peakAnno <- annotatePeak(macsPeaks_GR, tssRegion=c(-1000, 1000), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")
```

---
## Annotation of peaks to genes
This allowed us to produce a GRanges or data.frame of our peaks and their predicted target genes.

```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR <- as.GRanges(peakAnno)
annotatedPeaksDF <- as.data.frame(peakAnno)
annotatedPeaksDF[1:2,]
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gene Set Enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gene Set Enrichment

---
"    
  )
  
}

```


---
## Geneset testing

Transcription factors or epigenetic marks may act on specific sets of genes grouped by a common biological feature (shared Biological function, common regulation in RNAseq experiment etc). 

A frequent step in ChIPseq analysis is to test whether common gene sets are enriched for transcription factor binding or epigenetic marks. We performed similar analysis in our [RNAseq workflow](https://rockefelleruniversity.github.io/RU_RNAseq/presentations/slides/RU_RNAseq_p3.html#1) looking for enrichment of gene sets in differentially changing genes.

Sources of well curated genesets include [GO consortium](http://geneontology.org/) (gene's function, biological process and cellular localisation), [REACTOME](http://www.reactome.org/) (Biological Pathways) and [MsigDB](http://software.broadinstitute.org/gsea/msigdb/) (Computationally and Experimentally derived).

---
## Gene Ontology

The [Gene Ontology consortium]((http://geneontology.org/)) aims to provide a comprehensive resource of the currently available knowledge regarding the functions of genes and gene products.

Functional categories of genes are broadly split into three main groups.

* **Molecular functions.** - Activity of a gene's protein product.
* **Biological processes.** - Role of gene's protein product.
* **Cellular components.** - Where in cell molecular function of protein product is performed.

---
## Gene Ontology

The three sub categories of gene ontology are arranged in nested, structured graph with gene sets at the top of graph representing more general terms and those at the bottom more specific terms.

![](imgs/goGraph.png)


---
## Reactome and KEGG

The [Reactome](http://www.reactome.org/) and [KEGG (Kyoto Encyclopedia of genes and genomes)](http://www.genome.jp/kegg/kegg2.html) contains information on genes' membership and roles in  molecular pathways.

These databases focus largely on metabolic and disease pathways, and allow us to investigate our genes in the context of not only functional roles but relative positions within pathways.


<div align="center">
<img src="imgs/map01100.png" alt="offset" height="300" width="600">
</div>



---
## MsigDB.
.pull-left[
The [molecular signatures database (MsigDB)](http://software.broadinstitute.org/gsea/msigdb/) is available from the Broad institute and provides a set of curated gene sets derived from sources such as GO, pathway databases, motif scans and even other experimental sets.

MsigDB databases are widely used in gene set enrichments analysis and are available as plain text in formats used with the popular Java based gene set enrichment software GSEA.
]
.pull-right[
<div align="center">
<img src="imgs/msigdb.png" alt="offset" height="500" width="350">
</div>
]


---
## Genesets in Bioconductor

In R we can access information on these gene sets through database libraries (such as the Org.db.eg we have reviewed)  such as **GO.db**, **KEGG.db**, **reactome.db** or by making use of libraries which allows us to our gene sets from parse plain text formats, **GSEABase**.


```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE,tidy=T}
library(GO.db)
library(KEGG.db)
library(reactome.db)
library(GSEABase)
```

---
## Gene ontology and geneset testing for ChIPseq

Geneset enrichment testing may be performed on the sets of genes with peaks associated to them. In this example we will consider genes with peaks within 1000bp of a gene's TSS.

We will not access these database libraries directly in testing but will use other R/Bioconductor libraries which make extensive use of them.


<div align="center">
<img src="imgs/TSSPeak.png" alt="offset" height="250" width="600">
</div>

---
## Gene ontology and geneset testing

To perform geneset testing here, we will use the **GOseq package**.

We must provide a named numeric vector of 1s or 0s to illustrate whether a gene had any peaks overlapping it's TSS.

In this example we use all TSS sites we found to be overlapped by Myc peaks.

The peaks landing in TSS regions will be marked as "Promoter" in the **annotation** column of our annotated GRanges object. 

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR[1,]
```

---
## Gene ontology and geneset testing


We can extract the unique names of genes with peaks in their TSS by subsetting the annotated GRanges and retrieving gene names from the **geneId** column.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR_TSS <- annotatedPeaksGR[
  annotatedPeaksGR$annotation == "Promoter",]
genesWithPeakInTSS <- unique(annotatedPeaksGR_TSS$geneId)
genesWithPeakInTSS[1:2]
```

---
# Gene ontology and functional testing.


Next we can extract all genes which are included in the TxDb object to use as our universe of genes for pathway enrichment.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
allGeneGR <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
allGeneGR[1:3,]
allGeneIDs <- allGeneGR$gene_id
```

---
# Gene ontology and functional testing.


Once we have a vector of all genes we can create a named vector of 1s or 0s representing whether a gene had peak in TSS or not. 
We can turn a logical vector into 1 for TRUE and 0 for FALSE simply using the **as.integer()** function.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
allGenesForGOseq <- as.integer(allGeneIDs %in% genesWithPeakInTSS)
names(allGenesForGOseq) <- allGeneIDs
allGenesForGOseq[1:3]
```


---

# Gene ontology and functional testing.

Now we have the the input for GOseq we can test against KEGG (or GO if we choose) using a standard hypergeometric test.

First we must construct a **nullp** data.frame for use within **goseq** using the **nullp()** function and supplying our named vector, genome to be used and gene identifier used.

The **nullp()** function attempts to correct for gene length biases we may see in geneset testing. i.e. a longer gene may have more chance to have a peak within it.


```{r,include=FALSE}
library(goseq)
```

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(goseq)
pwf=nullp(allGenesForGOseq,"mm10","knownGene",plot.fit=FALSE)
```

---

# Gene ontology and functional testing.

We can see which genomes are supported using the **supportedGenomes()** function.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
supGenomes <- supportedGenomes()
supGenomes[1:2,]

```

---

# Gene ontology and functional testing.

Now we have created our nullp data.frame, we can use this within the **goseq()** function to test for the enrichment of genesets.

We also select a database of genesets to test, here "KEGG". We can choose additionally from "GO:CC", "GO:BP", "GO:MF" genesets.

As we are testing peaks in 1000bp windows around genes's TSSs we can simply use the **"Hypergeometric"** for the method parameter.

```{r,echo=T, eval=F, echo=T, warning=FALSE}

Kegg_MycPeaks <- goseq(pwf,"mm10","knownGene",
                       test.cats=c("KEGG"),
                       method="Hypergeometric")

Kegg_MycPeaks[1:2,]

```

```{r,include=FALSE}

Kegg_MycPeaks <- goseq(pwf,"mm10","knownGene",
                       test.cats=c("KEGG"),
                       method="Hypergeometric")

Kegg_MycPeaks[1:2,]

```

```{r,eval=T,echo=F, eval=T,warning=FALSE}
Kegg_MycPeaks[1:2,]

```
---
# Gene ontology and functional testing.

The dataframe of enrichment results contains KEGG IDs but no description of pathways and processes.

We can use the **KEGG.db** package to extract a mapping of IDs to descriptive statistics.
```{r,include=FALSE}
library(KEGG.db)
```

```{r}
library(KEGG.db)
xx <- as.list(KEGGPATHID2NAME)
idtoName <- cbind(names(xx),unlist(xx))
idtoName[1:5,]

```
---
# Gene ontology and functional testing.

We can now merge our data.frame of ID to pathway name mappings to our data.frame of enrichment results.

```{r}
KeggN_MycPeaks <- merge(idtoName,Kegg_MycPeaks,by=1,all=TRUE)
orderByP <- order(KeggN_MycPeaks$over_represented_pvalue)
KeggN_MycPeaks <- KeggN_MycPeaks[orderByP,]
KeggN_MycPeaks[1:5,]
```
---

# Gene ontology and functional testing. GREAT method.

In addition to a standard enrichment tests, methods have been implemented specifically for ChIPseq. Many of these tools aim to incorporate peaks distal to genes into their enrichment testing such as the popular [GREAT](http://bejerano.stanford.edu/great/public/html/splash.php) toolset.

Incorporating distal peaks by rules such as nearest gene results in some genes having a higher chance of being selected and hence some genesets as a whole having a higher chance of having its members selected.

[GREAT](http://bejerano.stanford.edu/great/public/html/splash.php) defines regulatory regions for each, individual gene and compares the proportion of peaks mapping to a geneset's regulatory regions to the proportion of the genome occupied by geneset's regulatory regions.

i.e. If a gene set's regulatory regions account for 1% of the genome then one might expect 1% of peaks to overlap these regions by chance.
---
rGREAT - R interface to GREAT server.
=========================================================

We can use the GREAT Bioconductor interface available in the rGREAT package. 

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(rGREAT)
```
---
Gene ontology and functional testing. GREAT method.
=========================================================

To submit jobs we can use our GRanges of Myc peaks and specify a genome with the **submitGreatJob** function.

This function returns a GreatJob object containing a reference to our results on the GREAT server. To review the categories of results available we can use the availableCategories function on our GreatJob object.

```{r,eval=T,echo=T, eval=T, echo=T,messages=F,message=F,warning=FALSE,tidy=T}
great_Job <- submitGreatJob(macsPeaks_GR,species="mm10")
availableCategories(great_Job)
```
---
Gene ontology and functional testing. GREAT method.
=========================================================

The results table can be retrieved using the getEnrichmentTables function and specifying which tables we wish to review.

Here we retrieve the results tables for the "Regulatory Motifs" genesets which contains 2 seperate database results.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
great_ResultTable = getEnrichmentTables(great_Job,
                                        category="Regulatory Motifs")
names(great_ResultTable)
```
---
Gene ontology and functional testing. GREAT method.
=========================================================

Now we can review the enrichment of our genes with Myc peaks in their TSS for the "MSigDB Predicted Promoter Motifs" genesets.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
msigProMotifs <- great_ResultTable[["MSigDB Predicted Promoter Motifs"]]
msigProMotifs[1:4,]
```

--- 

# Motifs

A common practice in transcription factor ChIPseq is to investigate the motifs enriched under peaks. 

Denovo motif enrichment can be performed in R/Bioconductor but this can be very time consuming. Here we will use the Meme-ChIP suite available online to identify denovo motifs.

Meme-ChIP requires a FASTA file of sequences under peaks as input so we extract this using the **BSgenome** package.

---

# Extracting sequences under regions


First we need to load the BSgenome object for the genome we are working on, UCSC's mm10 build for the mouse genome, **BSgenome.Mmusculus.UCSC.mm10**.

```{r, echo=TRUE,include=FALSE}

library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
BSgenome.Mmusculus.UCSC.mm10
```


```{r, echo=TRUE,collapse=F}

library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
BSgenome.Mmusculus.UCSC.mm10
```
---

#Extracting sequences under regions


The motif for the ChIP-ed transcription factor should in the centre of a peak. Meme-ChIP will trim our peaks to a common length internally if sequences are of different length.

It is best therefore to provide a peak set resized to a common length.

```{r, echo=TRUE,collapse=F}
macsSummits_GR <- GRanges(seqnames(macsPeaks_GR),
                          IRanges(macsPeaks_GR$abs_summit,
                                  macsPeaks_GR$abs_summit),
                          score=macsPeaks_GR$fold_enrichment)
macsSummits_GR <- resize(macsSummits_GR,100,fix="center")

```

---
#Extracting sequences under regions

We now have a GRanges, centred on the summit, highest point of signal for every peak.

```{r, echo=TRUE,collapse=F}
macsSummits_GR
```

---
#Extracting sequences under regions


Once we have recentered our peaks we can use the **getSeq** function with our GRanges of resized common peaks and the BSgenome object for mm10.

The **getSeq** function returns a *DNAStringSet* object containing sequences under peaks. 

```{r, echo=TRUE,collapse=F}
peaksSequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10,
                         macsSummits_GR)
names(peaksSequences) <- paste0(seqnames(macsSummits_GR),":",
                                         start(macsSummits_GR),
                                         "-",
                                         end(macsSummits_GR))

peaksSequences[1:2,]
```
---
# Writing to FASTA file


The *writeXStringSet* function allows the user to write DNA/RNA/AA(aminoacid)StringSet objects out to file. 

By default the *writeXStringSet* function writes the sequence information in FASTA format (as required for Meme-ChIP).

```{r, echo=TRUE,collapse=F}
writeXStringSet(peaksSequences,file="mycMel_rep1.fa")

```

---

Now the file "mycMel_rep1.fa" contains sequences around the geometric centre of peaks suitable for Motif analysis in Meme-ChIP. 

In your own work you will typically run this from your own laptop with Meme installed locally but today we will upload our generated FASTA file to their [web portal](http://meme-suite.org/tools/meme-chip). 

Results files from Meme-ChIP can be found [here](http://rockefelleruniversity.github.io/myc_Meme_Example/meme-chip.html)

---
# Parsing back fimo results

We can retrieve the locations of Myc motifs identified in Meme-ChIP from the Fimo output.

Fimo reports Myc motif locations as a GFF3 file which we should be able to visualise in IGV. 
Sadly, this GFF file's naming conventions cause only a fraction of motifs to be reported.

<div align="center">
<img src="imgs/fimoBad.png" alt="offset" height="300" width="600">
</div>

---
# Fimo to R.

Fortunately we can parse our motif's GFF file into R and address this using the **import()** function in  the **rtracklayer** package.

```{r, echo=TRUE,collapse=F}
library(rtracklayer)
motifGFF <- import("~/Downloads/fimo.gff")
```
---
# Fimo to valid GFF3.

We can give the sequences some more sensible names and export the GFF to file to visualise in IGV.

```{r, echo=TRUE,collapse=F}
motifGFF$Name <- paste0(seqnames(motifGFF),":",
                        start(motifGFF),"-",end(motifGFF))
motifGFF$ID <- paste0(seqnames(motifGFF),":",
                      start(motifGFF),"-",end(motifGFF))
export.gff3(motifGFF,con="~/Downloads/fimoUpdated.gff")
```

<div align="center">
<img src="imgs/fimoGood.png" alt="offset" height="250" width="600">
</div>


---
<!-- --- -->
<!-- # High confidence peaks -->

<!-- As we discussed in our last session, ChIPseq data will often contain artefact signal and noise. Following the work by the Encode consortium, it is becoming more standard practice to have replicate samples for ChIPseq. -->

<!-- One approach to selecting reproducible peaks in ChIPseq is to identify peaks which are present in the majority of replicates. -->

<!-- We can first read in the other replicate dataset for MEL Myc ChIPseq below. -->

<!-- ```{r, echo=TRUE,collapse=F} -->

<!-- library(GenomicRanges) -->
<!-- library(TxDb.Mmusculus.UCSC.mm10.knownGene) -->
<!-- library(ChIPseeker) -->
<!-- macsPeaks <- "data/peaks/Mel_2_peaks.xls" -->
<!-- macsPeaks_DF2 <- read.delim(macsPeaks,comment.char="#") -->
<!-- macsPeaks_GR2 <- GRanges( -->
<!--  seqnames=macsPeaks_DF2[,"chr"], -->
<!--  IRanges(macsPeaks_DF2[,"start"],macsPeaks_DF2[,"end"]) -->
<!-- ) -->
<!-- mcols(macsPeaks_GR2) <- macsPeaks_DF2[,c("abs_summit", "fold_enrichment")] -->
<!-- ``` -->

<!-- --- -->
<!-- # Manipulating Peak Sets - Finding Common peaks -->


<!-- When looking at peaks which occur in both samples it is clear that the number of peaks in first replicate overlapping those in second is different from number of second replicate peaks overlapping first. -->

<!-- This is because 2 peaks from one replicate may overlap 1 peak in the other replicate. -->

<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- firstANDsecondPeakSets <- macsPeaks_GR[macsPeaks_GR %over% macsPeaks_GR2] -->
<!-- secondANDfirstPeakSets <- macsPeaks_GR2[macsPeaks_GR2 %over% macsPeaks_GR] -->

<!-- length(firstANDsecondPeakSets) -->
<!-- length(secondANDfirstPeakSets) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/oneToMany.png) -->

<!-- --- -->
<!-- # Manipulating Peak Sets - Finding Common peaks -->


<!-- A common step with finding overlapping transcription factor peaks is to reduce peaksets to single non-overlapping peakset before interrogating whether a peak occurred in a sample. -->

<!-- This allows for a single peak set to be used as a consensus peakset between replicates. -->


<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- allPeaks <- c(macsPeaks_GR,macsPeaks_GR2) -->
<!-- allPeaksReduced <- reduce(allPeaks) -->
<!-- length(allPeaks) -->
<!-- length(allPeaksReduced) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/mel_Flattened.png) -->


<!-- --- -->

<!-- Now we can use a logical expression to subset our reduced/flattened peak set to those overlapping peaks in both the first and second replicate. -->

<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- commonPeaks <- allPeaksReduced[allPeaksReduced %over% macsPeaks_GR  -->
<!--                                & allPeaksReduced %over% macsPeaks_GR2] -->
<!-- length(commonPeaks) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/Ch12_highcon.png) -->

---

# Time for an exercise.

[Link_to_exercises](../../Exercises/chipseq_part3_exercises.html)

[Link_to_answers](../../Answers/chipseq_part3_answers.html)


