---
title: "ChIPseq In Bioconductor part4"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_ChIPseq/"
output: 
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


# Data

In todays session we will continue to review the Myc ChIPseq we were working on in our last sessions.

This include Myc ChIPseq for MEL and Ch12 celllines.

Information and files for the [Myc ChIPseq in MEL cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000EUA/)

Information and files for the [Myc ChIPseq in Ch12 cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000ERN/)

<!-- --- -->

<!-- # Data -->

<!-- We will be working with peak calls today, so we can download the MACS2 peak calls from the Encode website. -->

<!-- [Myc Mel Rep1](https://www.encodeproject.org/files/ENCFF363WUG/@@download/ENCFF363WUG.bed.gz) -->

<!-- [Myc Mel Rep2](https://www.encodeproject.org/files/ENCFF139JHS/@@download/ENCFF139JHS.bed.gz) -->

<!-- [Myc Ch12 Rep1](https://www.encodeproject.org/files/ENCFF160KXR/@@download/ENCFF160KXR.bed.gz) -->

<!-- [Myc Ch12 Rep2](https://www.encodeproject.org/files/ENCFF962BGJ/@@download/ENCFF962BGJ.bed.gz) -->

---

# Data

I have provided peakcalls from MACS2 following the processing steps outlined in our last session.

Peak calls for Myc in MEL and Ch12 cellines can be found in 

**Data/peaks/**

* **Data/peaks/Mel_1_peaks.xls**
* **Data/peaks/Mel_2_peaks.xls**
* **Data/peaks/Ch12_1_peaks.xls**
* **Data/peaks/Ch12_1_peaks.xls**

---
# Mapping Transcription factor binding and epigenetic states

A common goal in ChIPseq is characterise genome wide transcription factor binding sites or epigenetic states. 

The presence of transciption factor binding sites and epigenetics events is often further analysed in the context of their putative targets genes to characterise the transciption factor's and epigenetic event's function and/or biological role.

<div align="center">
<img src="imgs/singleMap.png" alt="offset" height="300" width="600">
</div>

---
# Mapping Transcription factor binding and epigenetic states

With the release of Encode's wide scale mapping of transcription factor binding sites or epigenetic states and the advent of multiplexing technologies for high throughput sequencing, in has become common practice to have replicated ChIPseq experiments so as to have higher confidence in identified epigenetic events.

<div align="center">
<img src="imgs/MappingEvents.png" alt="offset" height="300" width="600">
</div>


---
# Changes in Transcription factor binding and epigentic states.

In addition to the genome wide characterisation of epigenetic events, ChIPseq had been increasingly used to identify changes in epigenetic events between conditions and/or cell lines.

<div align="center">
<img src="imgs/hoxDiff2.png" alt="offset" height="300" width="600">
</div>


---
# Myc ChIPseq in Mel and Ch12 cell-lines.

We have been working to process and a characterise Myc ChIPseq replicate in the Mel cell-line to gain an understanding of a potential role for Myc in Mel cell-line.

In this session we will look at how we can define a high confidence/reproducible set of Myc peaks in the Mel cellline as well as identify Myc binding events unique or common between Mel and Ch12 cell-lines. 


---
# Reading in a set of peaks

First we need to read our peak calls from MACS2 into R. 

The Myc peak calls we will review are within the peaks directory, so here we list all files matching our expected file pattern using the **dir() function.**

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
peakFiles <- dir("../../Data/peaks/",pattern="*.peaks",
                 full.names = TRUE)
peakFiles

```


---
# Reading in a set of peaks

We can loop through our tab seperated files (disguised as *.xls* functions) and import them into R as a list of data.frames using a loop.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
macsPeaks_DF <- list()
for(i in 1:length(peakFiles)){
  macsPeaks_DF[[i]] <- read.delim(peakFiles[i],
                                  comment.char="#")
}
length(macsPeaks_DF)
```

---
# Reading in a set of peaks

Now with our list of data.frames of peak calls, we loop through the list and create a **GRanges** for each of our peak calls.
```{r, include=FALSE}
library(GenomicRanges)
library(Rsamtools)
library(rtracklayer)
library(GenomicAlignments)
library(tracktables)
library(limma)
```

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
library(GenomicRanges)
macsPeaks_GR <- list()
for(i in 1:length(macsPeaks_DF)){
     peakDFtemp <- macsPeaks_DF[[i]]
     macsPeaks_GR[[i]] <- GRanges(
     seqnames=peakDFtemp[,"chr"],
     IRanges(peakDFtemp[,"start"],
             peakDFtemp[,"end"]
     )
  )
}
macsPeaks_GR[[1]]
```

---
# Reading in a set of peaks

We will want to assign a sensible set of names to our peak calls.

We can use the **gsub()** and **basename()** function with our file names to create some samplenames.

The **basename()** function accepts a file path (such as the path to our bam files) and returns just the file name (removing directory paths).

The **gsub()** function accepts the text to replace, the replacement text and a character vector to replace within.



```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
fileNames <- basename(peakFiles)
fileNames
sampleNames <- gsub("_peaks.xls","",fileNames)
sampleNames
```

---
# Reading in a set of peaks

Now we have a named list of our peak calls as **GRanges** objects.

We can convert our list of **GRanges** objects to a **GRangesList** using the **GRangesList()** function.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
macsPeaks_GRL <- GRangesList(macsPeaks_GR)
names(macsPeaks_GRL) <- sampleNames
class(macsPeaks_GRL)
names(macsPeaks_GRL)
```

---
# GRangesList objects

The GRangesList object can behave just as our standard lists. Here we use the **lengths()** function to a get the number of peaks in each replicate.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
lengths(macsPeaks_GRL)
```


---
# GRangesList objects

A major advantage of **GRangesList** objects is that we can apply many of the **GRanges** accessor and operator functions directly to our **GRangesList**. 

This means there is no need to lapply and convert back to **GRangesList** if we wish to alter our **GRanges** by a common method.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
library(rtracklayer)
macsPeaks_GRLCentred <- resize(macsPeaks_GRL,10,fix="center")
width(macsPeaks_GRLCentred)
```

---
# GRangesList objects

Now we have our GRangesList we can extract the peak calls for the Mel replicates.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
Mel_1_Peaks <- macsPeaks_GRL$Mel_1
Mel_2_Peaks <- macsPeaks_GRL$Mel_2
length(Mel_1_Peaks)
length(Mel_2_Peaks)
```

---

# Finding unique peaks

We can extract peak calls unique to replicate 1 or 2 using the **%over%** operator.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
Mel_1_Unique <- Mel_1_Peaks[!Mel_1_Peaks %over% Mel_2_Peaks]
Mel_2_Unique <- Mel_2_Peaks[!Mel_2_Peaks %over% Mel_1_Peaks]
length(Mel_1_Unique)
length(Mel_2_Unique)
export.bed(Mel_1_Unique,"Mel_1_Unique.bed")
export.bed(Mel_2_Unique,"Mel_2_Unique.bed")
```

---

# Finding unique peaks

![](imgs/uniquePeaks.png)
---

# Finding common peaks

Similarly we can extract peak calls common to replicate 1 or 2.

The numbers in common however differ. This is because 2 peak calls in one sample can overlap 1 peak call in the other replicate.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
Mel_1_Common <- Mel_1_Peaks[Mel_1_Peaks %over% Mel_2_Peaks]
Mel_2_Common <- Mel_2_Peaks[Mel_2_Peaks %over% Mel_1_Peaks]
length(Mel_1_Common)
length(Mel_2_Common)
export.bed(Mel_1_Common,"Mel_1_Common.bed")
export.bed(Mel_2_Common,"Mel_2_Common.bed")
```

---

![](imgs/commonP.png)

---

![](imgs/2for1.png)

---
# Defining a consensus, redundant set.

To address this problem, a common operation in ChIPseq is to define a non-redundant set of peaks across all samples.

To do this we first pool all our peaks across all replicates, here Mel and Ch12, into one set of redundant, overlapping peaks.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
allPeaksSet_Overlapping <- unlist(macsPeaks_GRL)
allPeaksSet_Overlapping
```

---
# Defining a consensus, non-redundant set.

We can then use the **reduce()** function to collapse our peaks into non-redundant, distinct peaks representing peaks present in any sample.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
allPeaksSet_nR <- reduce(allPeaksSet_Overlapping)
allPeaksSet_nR
export.bed(allPeaksSet_nR,"allPeaksSet_nR.bed")
```

---
![](imgs/nrp.png)

---
# Defining a common set of peaks.

With our newly defined non-redundant peak set we can now identify from this set which peaks were present in both our replicates using the **%over%** operator and a logical expression.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
commonPeaks <- allPeaksSet_nR[allPeaksSet_nR %over% Mel_1_Peaks &
                               allPeaksSet_nR %over% Mel_2_Peaks]
commonPeaks
export.bed(commonPeaks,"commonPeaks.bed")

```

---

![](imgs/commonPR.png)
---
# Defining a unique set of peaks.

Similarly we can identify which peaks are present only in one replicate.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
mel1_Only <- allPeaksSet_nR[allPeaksSet_nR %over% Mel_1_Peaks &
                             !allPeaksSet_nR %over% Mel_2_Peaks]
mel2_Only <- allPeaksSet_nR[!allPeaksSet_nR %over% Mel_1_Peaks &
                             allPeaksSet_nR %over% Mel_2_Peaks]
length(mel1_Only)
length(mel2_Only)
export.bed(mel1_Only,"mel1_Only.bed")
export.bed(mel2_Only,"mel2_Only.bed")

```

---
![](imgs/newMelOnly.png)
---
# Complex overlaps.

When working with larger numbers of peaks we will often define a logical matrix describing in which samples our non-redundant peaks were present.

First then we use a loop to generate a logical vector for the occurence of non-redundant peaks in each sample.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
overlap <- list()
for(i in 1:length(macsPeaks_GRL)){
  overlap[[i]] <- allPeaksSet_nR %over% macsPeaks_GRL[[i]]
}
overlap[[1]][1:2]
```

---
# Complex overlaps.

We can now use to **do.call** and **cbind** function to column bind our list of overlaps into our matrix of peak occurrence.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
overlapMatrix <- do.call(cbind,overlap)
colnames(overlapMatrix) <- names(macsPeaks_GRL)
overlapMatrix[1:2,]
```

---
# Complex overlaps.

We can add the matrix back into the metadata columns of our **GRanges()** of non-redundant peaks using the **mcols()** accessor.

Now we have our non-redundant peaks and the occurence of these peaks in every sample we can easily identify peaks unique or common to replicates and conditions/cell-lines.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
mcols(allPeaksSet_nR) <- overlapMatrix
allPeaksSet_nR[1:2,]
```

---
# Complex overlaps.

The **limma** package is commonly used in the analysis of RNAseq and microarray data and contains many useful helpfull functions.

One very useful function is the **vennDiagram** function which allows us to plot overlaps from a logical matrix, just like the one we created.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,fig.height=5,fig.width=5}
library(limma)
vennDiagram(mcols(allPeaksSet_nR))
```

---
# Complex overlaps.

The **limma** package's **vennCounts** function allows us to retrieve the counts displayed in the Venn diagram as a data.frame.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
vennCounts(mcols(allPeaksSet_nR))
```

---
# High confidence peaks.

With our non-redundant set of peaks and our matrix of peak occurrence, we can define replicated peaks within conditions.

Here we define the peaks which occur in both the Ch12 replicates.

Since logical matrix is equivalent to a 1 or 0  matrix (1 = TRUE and 0 = FALSE), we can use the rowSums function to extract peaks in at least 2 of the Ch12 replicates.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
ch12_HC_Peaks <- allPeaksSet_nR[
  rowSums(as.data.frame(
    mcols(allPeaksSet_nR)[,c("ch12_1","ch12_2")])) >= 2
  ]
export.bed(ch12_HC_Peaks,"ch12_HC_Peaks.bed")
ch12_HC_Peaks[1:2,]
```

---
![](imgs/ch12HC.png)
---
# High confidence unique peaks.

Similarly we can define peaks which are replicated in Ch12 but absent in Mel samples.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
ch12_HC_UniquePeaks <- allPeaksSet_nR[
  rowSums(as.data.frame(
    mcols(allPeaksSet_nR)[,c("ch12_1","ch12_2")])) >= 2 &
  rowSums(as.data.frame(
    mcols(allPeaksSet_nR)[,c("Mel_1","Mel_2")])) == 0  
  ]
export.bed(ch12_HC_UniquePeaks,"ch12_HC_UniquePeaks.bed")
ch12_HC_UniquePeaks[1,]
```

---
![](imgs/ch12U.png)

---
# Finding differential regions.

Identifying peaks specific to cell-lines or conditions however does not capture the full range of changes in epigenetic events.

To identify differences in epigenetic events we can attempty to quantify the changes in fragment abundance from IP samples across our non-redundant set of peaks.

![](imgs/igsss_1.png)

---
# Finding differential regions.

We first must establish a set of regions within which to quantify IP-ed fragments.

An establised technique is to produce a set of non-redundant peaks which occur in the majority of at least one experimental condition under evaluation.

Here we identify peaks which occurred in both replicates in either Mel or Ch12 cell-lines.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
HC_Peaks <- allPeaksSet_nR[
  rowSums(as.data.frame(
    mcols(allPeaksSet_nR)[,c("ch12_1","ch12_2")])) >= 2 |
  rowSums(as.data.frame(
    mcols(allPeaksSet_nR)[,c("Mel_1","Mel_2")])) >= 2  
  ]
HC_Peaks
export.bed(HC_Peaks,"HC_Peaks.bed")
```

---
![](imgs/hcPeaks2.png)
---
# Counting regions.

We will count from our aligned BAM files to quantify IP fragments.

As we have seen previously we can use the **BamFileList()** function to specify which BAMs to count and importantly, to control memory we specify the number of reads to be held in memory at one time using the **yield()** parameter.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}

library(Rsamtools)

bams <- c("~/Projects/Results/chipseq/testRun/BAMs/Sorted_Myc_Ch12_1.bam",
          "~/Projects/Results/chipseq/testRun/BAMs/Sorted_Myc_Ch12_2.bam",
          "~/Projects/Results/chipseq/testRun/BAMs/Sorted_Myc_Mel_1.bam",
          "~/Projects/Results/chipseq/testRun/BAMs/Sorted_Myc_Mel_2.bam")
bamFL <- BamFileList(bams,yieldSize = 5000000)
bamFL
```


---
# Counting regions.

We can count the number of fragments overlapping our peaks using the **summarizeOverlaps** function. Since ChIPseq is strandless, we set the **ignore.strand** parameter to **TRUE**.

The returned object is a familiar **RangedSummarizedExperiment** containing our GRanges of non-redundant peaks and the counts in these regions for our BAM files.

```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE}
library(GenomicAlignments)
myMycCounts <- summarizeOverlaps(HC_Peaks,
                              reads = bamFL,
                              ignore.strand = TRUE)
class(myMycCounts)
save(myMycCounts,file="../../Data/MycCounts.RData")
```



```{r, eval=T, echo=F, warning=FALSE}
suppressPackageStartupMessages(library(GenomicAlignments))
load("../../Data/MycCounts.RData")
class(myMycCounts)
```

---
# Differential regions using DESeq2

To assess changes in ChIPseq signal across cell-lines we will use the **DESeq2** package.

The DESeq2 package contains a workflow for assessing local changes in fragment/read abundance between replicated conditions. This workflow includes normalisation, variance estimation, outlier removal/replacement as well as significance testing suited to high throughput sequencing data (ie. integer counts).

To make use of DESeq2 workflow we must first create a data.frame of conditions of interest with rownames set as our BAM file names.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
metaDataFrame <- data.frame(CellLine=c("Ch12","Ch12","Mel","Mel"))
rownames(metaDataFrame) <- colnames(myMycCounts)
metaDataFrame
```

---
# Differential regions using DESeq2

We can use the **DESeqDataSetFromMatrix()** function to create a **DESeq2** object.

We must provide our matrix of counts to **countData** parameter, our metadata data.frame to **colData** parameter and we include to an optional parameter of **rowRanges** the non-redundant peak set we can counted on.

Finally we provide the name of the column in our metadata data.frame within which we wish to test to the **design** parameter.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
library(DESeq2)
deseqMyc <- DESeqDataSetFromMatrix(countData = assay(myMycCounts),
                              colData = metaDataFrame,
                              design = ~ CellLine,
                              rowRanges=HC_Peaks)
```

---
# Differential regions using DESeq2

We can now run the DESeq2 workflow on our **DESeq2** object using the **DESeq()** function.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
deseqMyc <- DESeq(deseqMyc)
```

---
# Differential regions using DESeq2

Our **DESeq2** object is updated to include useful statistics such our normalised values and variance of signal within each non-redundant peak call.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
deseqMyc
```
---
# Differential regions using DESeq2

We can extract our information of differential regions using the **results()** function.

We provide to the **results()** function the **DESeq2** object, the comparison of interest to the **contrast** parameter and the type of output to return to the **format** parameter.

The comparison to **contrast** parameter is provided as a vector of length3 including the metadata column of interest and groups to test.

We can sort the results by pvalue using the **order()** function to rank by the most significant changes.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
MelMinusCh12 <- results(deseqMyc,
                        contrast = c("CellLine","Mel","Ch12"),
                        format="GRanges")
MelMinusCh12 <- MelMinusCh12[order(MelMinusCh12$pvalue),]
class(MelMinusCh12)
```

---
# Differential regions using DESeq2

The GRanges object contains information on the comparison made in DESeq2.

Most useful it contains the the difference in IP signal as log2 fold change in **log2FoldChange**, the significance of the change in the **pvalue** column and an adjusted p-value to address multiple correction in **padj** column.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
MelMinusCh12[1,]
```


---
# Differential regions using DESeq2

We can now filter our non-redundant peaks to those with significantly more signal in Mel or Ch12 cell-lines by filtering by log2FoldChange and padj (p-value adjusted for multiple correction) less than 0.05.


```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE}
MelMinusCh12Filt <- MelMinusCh12[!is.na(MelMinusCh12$pvalue) | !is.na(MelMinusCh12$padj)]
UpinMel <-  MelMinusCh12[MelMinusCh12$padj < 0.05 & MelMinusCh12$log2FoldChange > 0]
DowninMel <-  MelMinusCh12[MelMinusCh12$padj < 0.05 & MelMinusCh12$log2FoldChange < 0]
export.bed(UpinMel,"UpinMel.bed")
export.bed(DowninMel,"DowninMel.bed")
```


---
![](imgs/DB.png)


---
# Differential regions using DESeq2

Finally we can make our reviewing of sites in IGV alittle easier using the **tracktables** package. 

The **tracktables** package's **makebedtable()** function accepts a **GRanges** object and writes an HTML report contains links to IGV.

An example can be found [here](MelMinusCh12.html)

```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE}
library(tracktables)
myReport <- makebedtable(MelMinusCh12Filt,"MelMinusCh12.html",
                         basedirectory = getwd())

browseURL(myReport)
```

---

# Time for an exercise.

[Link_to_exercises](../../Exercises/chipseq_part4_exercises.html)

[Link_to_answers](../../Answers/chipseq_part4_answers.html)


<!-- --- -->
<!-- # Defining a consensus, non-redundant set. -->


<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- allPeaksSet_nR <- reduce(allPeaksSet_Overlapping,with.revmap=TRUE) -->
<!-- allPeaksSet_nR -->
<!-- ``` -->

